cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/io/cdap/cdap/datapipeline/StageOperationsValidator.java

```diff
@@ -163,34 +171,23 @@ void validate() {
       String field = next.getKey();
       List<FieldOperation> origins = next.getValue();
 
-      if (stageOutputs.contains(field)) {
-        if (origins.size() > 1) {
-          // size will be greater than 1 only if its redundant
-          origins.remove(origins.size() - 1);
-        } else {
-          iterator.remove();
-        }
-      } else {
-        // If this field is not in the output schema of the stage, it is valid.
-        // For example, a Joiner joins two datasets D1,D2 based on the joiner key D1.K1, D2.K2, and
-        // decides to drop the joiner key in the output schema. The operation
-        // [D1.K1, D2.K2] ->[K1, K2] is a valid even though K1,K2 are not in the output schema.
-
-        // this means the field has redundant, TODO: CDAP-15785 revist redundant validation
-        if (origins.size() > 1) {
-          continue;
-        }
-
-        // Origin is empty means this is an generated field, it can be considered valid.
-        // If the size is 1, this field has no redundant operations, it is valid.
-        iterator.remove();
+      if (origins.size() > 1) {
+        List<FieldOperation> operations = redundantOutputs.computeIfAbsent(field, k -> new ArrayList<>());
+        // except the last origin, all others are redundant
+        operations.addAll(origins.subList(0, origins.size() - 1));
       }
+
+      // No matter this field is or is not in the output schema of the stage, it is valid.
+      // For example, a Joiner joins two datasets D1,D2 based on the joiner key D1.K1, D2.K2, and
+      // decides to drop the joiner key in the output schema. The operation
+      // [D1.K1, D2.K2] ->[K1, K2] is a valid even though K1,K2 are not in the output schema.
+      iterator.remove();
     }
 
     this.invalidOutputs.putAll(unusedOutputs.entrySet().stream().collect(
       Collectors.toMap(Map.Entry::getKey,
                        e -> e.getValue().stream().map(FieldOperation::getName).collect(Collectors.toList()))));
-    this.invalidOutputs.putAll(redundantOutputs.entrySet().stream().collect(
+    this.redundantOutputs.putAll(redundantOutputs.entrySet().stream().collect(
```

Comment:
would it be a good idea to create another method for this since it uses exactly same logic?

Reply:
This is just a one line which does the convert. I don't think there is a need to create another method for it. And we may revisit the validation logic later since the current validation is too strict so the these fields might get changed in the future.